On response headers

pawelmhm opened the pull request:
Adds support for Lua callback `on_response_headers`. Callback will accept response object as argument, for now this object will have two methods: `get_header` and `abort()`. Using callback will allow users to abort requests if some header is present in response (e.g. content_type). 

https://github.com/scrapinghub/splash/issues/231

kmike commented on splash/network_manager.py:
```diff
@@ -266,9 +268,22 @@ def _handleFinished(self):
         self.log("Finished downloading {url}", reply)
 
     def _handleMetaData(self):
+        """Signal emitted before response body is reader, after
```
ready?

kmike commented:
WrappedRequest has one more feature - it disallows user to use itself outside of the callback (and breaks a GC cycle by removing a reference to LuaRuntime). Users can copy necessary request parameters if they need them. 

Do you think WrappedResponse shouldn't do it? There are two parts of the question:

1) we must ensure objects are not kept in memory longer than needed;
2) we should decide if using response outside callbacks is fine. `response.abort()` won't necessarily work outside a callback (response may be already downloaded), and keeping them for longer could lead to increased memory usage.

It could make sense to add _WrappedResponse to https://github.com/scrapinghub/splash/blob/aec50a784e3ccf920c3acfa1e5ce26d273da4887/splash/utils.py#L48 - this way _debug endpoint will be able to find alive _WrappedResponse objects.

kmike commented on splash/qtrender_lua.py:
```diff
@@ -723,6 +737,38 @@ def set_header(self, name, value):
+    def __init__(self, lua, response):
+        self.lua = lua
+        self.response = response
+        self.headers = dict([(str(k), str(v)) for k, v in response.rawHeaderPairs()])
```
Splash requires Python 2.7+, so it can be written more succinctly: 
`{str(k): str(v) for k,v in response.rawHeaderPairs()}`

While header->name mapping is convenient, it is not always correct - there could be several response headers with the same name, it is allowed by HTTP (see e.g. http://stackoverflow.com/questions/4371328/are-duplicate-http-response-headers-acceptable).

pawelmhm replied:
> While header->name mapping is convenient, it is not always correct 

this is valid point, but I see that PyQt4 QNetworkReply already handles this for us, if there are multiple headers with same key PyQT concatenates them together, you can test this like this

```python
def test_multiple_response_headers(self):
        resp = self.request_lua("""
        function main(splash)
            local h = nil
            splash:on_response_headers(function(response)
                h = response.get_headers()
            end)
            splash:http_get(splash.args.url)
            return h
        end
        """, {'url': "http://httpbin.org/response-headers?Set-Cookie=foo&Set-Cookie=bar&Set-Cookie=forest"})
```

above should return
```
{u'Access-Control-Allow-Credentials': u'true',
 u'Access-Control-Allow-Origin': u'*',
 u'Connection': u'keep-alive',
 u'Content-Length': u'130',
 u'Content-Type': u'application/json',
 u'Date': u'Tue, 07 Jul 2015 13:09:46 GMT',
 u'Server': u'nginx',
 u'Set-Cookie': u'foo\nbar\nforest'}
```
I see chrome devtools show multiple headers for this url:
```
Set-Cookie:foo
Set-Cookie:bar
Set-Cookie:forest
```

but they are concatenated by PyQt

kmike replied:
That's actually great. Could you please add a test for it?

kmike commented:
Hm, is there a way to get all response headers? To use response:get_header users need to know header names.

pawelmhm commented:
> Hm, is there a way to get all response headers? To use response:get_header users need to know header names.

ATM no, but this is easy to add  and definitely worth adding so will add that to this PR.

kmike commented:
This looks good! Could you please add

1) docs;
2) a test for `response:get_headers()`;
3) a test for multiple headers with the same name.

pawelmhm commented:
> 3) a test for multiple headers with the same name.

actually this is surprisingly difficult to test. We use twisted for mock resources but it looks like twisted does not allow to set multiple headers on response. [`request.setHeader(k, v)`](https://twistedmatrix.com/documents/current/api/twisted.web.http.Request.html#setHeader) keeps headers as dictionary, so if you try to set many headers with same key only last one will end up in response. To actually set multiple headers we could either find a way to do this with twisted mock resource (maybe somehow write raw response directly, but how?), or use remote resource for example httpbin.

pawelmhm commented:
hey @kmike can you check recent changes, I updated getting headers, removed method get_headers() and just added headers attribute to wrapped lua object making it available in lua script as a table (I think this should be easier for users). Also added basic response attributes: url, status and har info. If you think all ok I'll add docs.

kmike commented on splash/lua_modules/splash.lua:
```diff
@@ -223,5 +223,42 @@ function Splash:on_request(cb)
+
+    for key, value in pairs(py_reply.headers) do 
+        self[key] = value
+    end
```
Does it mean that if there is 'url' http header is the response user won't be able to access response.url attribute?

pawelmhm replied:
good catch looks like my mistake, will fix that

kmike commented:
The PR looks good, please go on :)

pawelmhm commented:
ok added docs and added request in callback @kmike

kmike commented on docs/scripting-ref.rst:
```diff
@@ -1560,6 +1560,80 @@ request to Splash:
+.. _splash-on-response-headers:
+
+splash:on_response_headers
+-----------------
```
the underline is too short, it won't render properly

kmike commented on splash/lua_modules/splash.lua:
```diff
@@ -223,5 +223,52 @@ function Splash:on_request(cb)
+    setmetatable(self, Response)
+    
+    -- convert har headers to something more convenient
+    _request_headers = {}
```
Shouldn't it be local?

kmike commented on docs/scripting-ref.rst:
```diff
@@ -1560,6 +1560,80 @@ request to Splash:
+* ``url`` - requested URL;
+* ``headers`` - HTTP headers of response
+* ``info`` - a table with response data in `HAR response`_ format
+* ``request`` - a table with request information 
```
I think we should document what's available in response.request table.

kmike commented on splash/lua_modules/splash.lua:
```diff
@@ -223,5 +223,52 @@ function Splash:on_request(cb)
+    for name, value in pairs(py_reply.request["headers"]) do
+        _request_headers[value["name"]] = value["value"]
+    end
+    py_reply.request["headers"] = _request_headers
```
Maybe we should provide this table for `splash:on_request` as well? It may be a separate ticket though, because for `splash:on_request` we need to make `request.headers` writeable.

kmike commented:
@pawelmhm there are some minor comments, but the PR looks solid :+1:
